frontend/src/store/auth/index.js

```diff
@@ -0,0 +1,103 @@
+import firebase from 'firebase/app'
+import 'firebase/auth'
+
+import config from './config'
+
+firebase.initializeApp(config)
+
+export const auth = firebase.auth()
+
+const state = {
+  loading: false,
+  error: '',
+  success: '',
+  user: null
+}
+
+const getters = {
+  loading: ({ loading }) => loading,
+  isLoggedIn: ({ user }) => !!user,
+  error: ({ error }) => error,
+  user: ({ user }) => user
+}
+
+setInterval(() => {
+  console.log('Background Auth state check: ', !!auth.currentUser)
+}, 10000)
+
+const mutations = {
+  ERROR(state, message) {
+    console.log('ERROR', message)
+    state.error = message
+  },
+  SIGN_UP(state, user) {
+    // on successful signup communicate with backend to update database.
+    // If by email, new users are signed in by default
+    console.log('SIGN UP', user)
+    state.success = 'Your account has been successfully created!'
+    state.user = user
+  },
+  LOGIN(state, user) {
+    // login happens in indexedDB in chrome, localStorage in other browsers
+    console.log('LOGIN', user)
+    state.user = user
+  },
+  LOGOUT(state) {
+    state.user = null
+    console.log('LOGOUT EVENT')
+  },
+  UPDATE(state, user) {
+    // useful for intial boot
+    console.log(user)
+    state.user = user
+  }
+}
+
+const actions = {
+  signIn({commit}, { email, password }) {
+    if (!(email && password)) {
+      return commit('ERROR', 'Missing fields')
+    }
+
+    return auth.signInWithEmailAndPassword(email, password)
+      .then(user => commit('LOGIN', user))
+      .catch(error => commit('ERROR', error.message))
+  },
+
+  logout({commit}) {
+    return auth.signOut()
+      .then(() => commit('LOGOUT'))
+      .catch(error => commit('ERROR', error.message))
+  },
+
+  signUp({commit}, { email, password }) {
+    if (!(email && password)) {
+      return commit('ERROR', 'Missing fields')
+    }
+
+    return auth.createUserWithEmailAndPassword(email, password)
+      .then(user => commit('SIGN_UP', user))
+      .catch(error => commit('ERROR', error.message))
+  },
+
+  checkAuth({commit}) {
+    return new Promise((resolve, reject) => {
+      const unsubscribe = auth.onAuthStateChanged(user => {
+        unsubscribe()
```

Comment:
da heck is this doing?

Reply:
??

Reply:
When the app boots, auth is still booting and interacting with the indexedDB. THis means this callback doesn't fire until it's ready. May be able to change it eventually. This seemed the least bad solution.

A related problem is that I think user should be in root state, so that the JWT can be sent in all requests if necessary.